### PR TITLE
create dynamic health tips

### PR DESCRIPTION
**Pull request by Baalmart**
**_WHAT DOES THIS PR DO?_**

- [x] create/manage dynamic health tips
- [x] Update the KAFKA `TIPS_TOPIC=tips-topic` on the KAFKA cluster configuration file
- [x] return all recent measurements with health tips

**_WHAT ISSUES ARE RELATED TO THIS PR?_**

- Jira cards
    - [MOB-164]

**_HOW DO I TEST OUT THIS PR?_**
```
cd src/device-registry
npm install
npm run stage-mac

```
**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [x] [get Events V1](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-2f1360d4-2eaa-40b7-9ad4-eef1b03abbce)
- [x] [get health tips](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-78323f45-60ca-45f1-ace2-6a43fad9761e)
- [x] [delete health tip](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-c8b4cdfa-fdd6-4e41-abd7-ce68b7720465)
- [x] [update health tip](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-49f9d6c0-48ca-4ead-952b-e6532c71a4de)
- [x] [create health tip](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-01e4dcca-8244-4264-b1f0-2ebbcd468946)

**_SCREENSHOTS?:_** 
Screenshot 2023-03-22 at 21 23 08
Screenshot 2023-03-23 at 07 55 05
Screenshot 2023-03-23 at 07 57 56
Screenshot 2023-03-23 at 08 00 30
Screenshot 2023-03-23 at 08 00 30
Screenshot 2023-03-23 at 08 17 20

[MOB-164]: https://airqoteam.atlassian.net/browse/MOB-164?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/device-registry/app.js [hunk 1]
```diff
@@ -51,6 +51,7 @@ app.use("/api/v1/devices/sites", routes.v1.sites);
 app.use("/api/v1/devices/events", routes.v1.events);
 app.use("/api/v1/devices/locations", routes.v1.locations);
 app.use("/api/v1/devices/photos", routes.v1.photos);
+app.use("/api/v1/devices/tips", routes.v1.tips);
```
Comment (Baalmart) [inline src/device-registry/app.js:54]: new routes for health tips

File: src/device-registry/controllers/create-health-tips.js [hunk 2]
```diff
@@ -0,0 +1,287 @@
+const HTTPStatus = require("http-status");
+const { logObject, logElement, logText } = require("@utils/log");
+const errors = require("@utils/errors");
+const constants = require("@config/constants");
+const log4js = require("log4js");
```
Comment (Baalmart) [inline src/device-registry/controllers/create-health-tips.js:5]: controller logic for health tips

File: src/device-registry/models/Event.js [hunk 3]
```diff
@@ -563,10 +563,21 @@ eventSchema.statics = {
...
+        /**
+         * look up the health tips and attach them to
+         * measurements as you return each one of them
+         * what is the direct relationship between measurement and pm2.5?
+         */
+        const data = await this.aggregate()
           .unwind("values")
           .match(search)
           .replaceRoot("values")
+          .lookup({
```
Comment (Baalmart) [inline src/device-registry/models/Event.js:575]: where the look up for health tips, using the pm2.5 value of measurements is supposed to happen from here. @BenjaminSsempala

File: src/device-registry/models/HealthTips.js [hunk 4]
```diff
@@ -0,0 +1,286 @@
+const { Schema, model } = require("mongoose");
+const uniqueValidator = require("mongoose-unique-validator");
+const mongoose = require("mongoose");
+const ObjectId = Schema.Types.ObjectId;
+const { logElement, logObject, logText } = require("@utils/log");
```
Comment (Baalmart) [inline src/device-registry/models/HealthTips.js:5]: the model/schema for health tips

File: src/device-registry/routes/v1/tips.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
+const express = require("express");
+const router = express.Router();
+const healthTipController = require("@controllers/create-health-tips");
+const { check, oneOf, query, body, param } = require("express-validator");
```
Comment (Baalmart) [inline src/device-registry/routes/v1/tips.js:4]: routes and corresponding input validations for v1 of health tips

File: src/device-registry/routes/v2/tips.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
+const express = require("express");
+const router = express.Router();
+const healthTipController = require("@controllers/create-health-tips");
+const { check, oneOf, query, body, param } = require("express-validator");
```
Comment (Baalmart) [inline src/device-registry/routes/v2/tips.js:4]: routes and corresponding input validations for v2 of health tips

File: src/device-registry/utils/create-health-tips.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
+const HTTPStatus = require("http-status");
+const HealthTipSchema = require("@models/HealthTips");
+const { getModelByTenant } = require("./multitenancy");
```
Comment (Baalmart) [inline src/device-registry/utils/create-health-tips.js:3]: utils for the health tips

File: k8s/kafka/topics/kafka-topics.yaml [hunk 8]
```diff
@@ -164,3 +164,17 @@ spec:
   replicas: 2
   config:
     retention.ms: 86400000
+
+---
```
Comment (Baalmart) [inline k8s/kafka/topics/kafka-topics.yaml:168]: add the health tips to the Kafka configuration file. @BenjaminSsempala

File: src/device-registry/utils/generate-filter.js [hunk 9]
```diff
@@ -904,6 +904,15 @@ const generateFilter = {
 
     return filter;
   },
+
+  tips: (request) => {
```
Comment (Baalmart) [inline src/device-registry/utils/generate-filter.js:908]: filter util function for health tips. We can also be able to specify other fields for querying tips using air quality measurements. @BenjaminSsempala

File: src/device-registry/models/HealthTips.js [hunk 10]
```diff
@@ -290,15 +290,15 @@ tipsSchema.statics = {
           description: 1,
         },
       };
-      const removedPhoto = await this.findOneAndRemove(filter, options).exec();
-      if (!isEmpty(removedPhoto)) {
+      const removedTip = await this.findOneAndRemove(filter, options).exec();
```
Comment (BenjaminSsempala) [inline src/device-registry/models/HealthTips.js:293]: @Baalmart Just renaming this to tip

**Pull request merged**